### PR TITLE
Add a shader validation cache to shader validation

### PR DESCRIPTION
**Pull request by Tony-LunarG**
Skips shader validation if the shader has already been validated and is in the validation cache.

The disable of this feature is a little messy.  I wanted it to be consistent with other disables, but didn't see a need to get it put into the spec.  If others find it too messy, I can start the process of adding it to the spec.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -2533,13 +2533,94 @@ void CoreChecks::PostCallRecordCreateDevice(VkPhysicalDevice gpu, const VkDevice
...
             core_checks->SetImageViewInitialLayout(cb_node, iv_state, layout);
         });
+
+    // Allocate shader validation cache
+    if (!disabled[shader_validation_caching] && !core_checks->coreValidationCache) {
+        std::string validation_cache_path;
+        auto tmp_path = GetEnvironment("TMPDIR");
+        if (!tmp_path.size()) tmp_path = GetEnvironment("TMP");
+        if (!tmp_path.size()) tmp_path = GetEnvironment("TEMP");
+        if (!tmp_path.size()) tmp_path = "//tmp";
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:2543]: Maybe worth creating a utility for? e.g., I think Windows has a dedicated API for this (`GetTempPath`?), whereas *nix uses environment variables...I think?

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -2533,13 +2533,94 @@ void CoreChecks::PostCallRecordCreateDevice(VkPhysicalDevice gpu, const VkDevice
...
+        void *validation_cache_data = nullptr;
+        FILE *read_file = fopen(core_checks->validation_cache_path.c_str(), "rb");
+
+        if (read_file) {
+            // Determine cache size
+            fseek(read_file, 0, SEEK_END);
+            validation_cache_size = ftell(read_file);
+            rewind(read_file);
+
+            validation_cache_data = (char *)malloc(sizeof(char) * validation_cache_size);
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:2556]: `std::vector`, `std::unique_ptr`? I'm not sure what the policy is on streams here, but you should be able to read the stream directly into a `std::vector`, though streams may only make more sense if we're using exceptions.

File: layers/core_validation.h [hunk 3]
```diff
@@ -110,6 +110,8 @@ class CoreChecks : public ValidationStateTracker {
     GlobalQFOTransferBarrierMap<QFOImageTransferBarrier> qfo_release_image_barrier_map;
     GlobalQFOTransferBarrierMap<QFOBufferTransferBarrier> qfo_release_buffer_barrier_map;
     GlobalImageLayoutMap imageLayoutMap;
+    VkValidationCacheEXT coreValidationCache = VK_NULL_HANDLE;
```
Comment (ncesario-lunarg) [inline layers/core_validation.h:113]: snake_case?

**Review by ncesario-lunarg — COMMENTED** [review]
Some minor nits and questions, but I think this looks good. I don't have a good sense as to the effects the "Disable2" API has. It looks like this is more or less transparent to end users?

File: layers/core_validation.cpp (hunk 2 above)
Comment (Tony-LunarG) [reply layers/core_validation.cpp]: Changed to ifstream->vector

File: layers/core_validation.h (hunk 3 above)
Comment (Tony-LunarG) [reply layers/core_validation.h]: Done

**Comment by Tony-LunarG** [comment]
I'll be pursuing adding the disable enum to the VkValidationFeatureDisableEXT enums

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -2533,13 +2534,81 @@ void CoreChecks::PostCallRecordCreateDevice(VkPhysicalDevice gpu, const VkDevice
...
+
+        if (read_file) {
+            // Determine cache size
+            std::streampos fileSize;
+            read_file.seekg(0, std::ios::end);
+            fileSize = read_file.tellg();
+            read_file.seekg(0, std::ios::beg);
+
+            validation_cache_data.reserve(static_cast<size_t>(fileSize));
+            validation_cache_data.assign(std::istreambuf_iterator<char>(read_file), std::istreambuf_iterator<char>());
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:2558]: :+1: this looks cleaner to me, but C-folks may disagree. "raw malloc" scares me...

I'm not sure if it's faster/"more efficient," but I think you can get rid of the file size query with `std::copy(std::istreambuf_iterator<char>(read_file), {}, std::back_inserter(validation_cache_data))`.

**Review by ncesario-lunarg — APPROVED** [review]

File: layers/core_validation.cpp (hunk 4 above)
Comment (Tony-LunarG) [reply layers/core_validation.cpp:2558]: Very cool - thanks!

File: layers/json/VkLayer_khronos_validation.json.in [hunk 5]
```diff
@@ -370,6 +370,12 @@
                             "label": "Idle Descritor Set",
                             "description": "",
                             "view": "ADVANCED"
+                        },
+                        {
+                            "key": "VALIDATION_FEATURE_DISABLE_SHADER_VALIDATION_CACHING_EXT",
```
Comment (christophe-lunarg) [inline layers/json/VkLayer_khronos_validation.json.in:375]: I would personally make it a bool. 

These "disables" settings are supposed to handle disable validation areas of the core validation.

That flag still seems orthogonal to me in term of feature.

**Review by ncesario-lunarg — APPROVED** [review]
